domain/agentprovisioner/service/service.go

```diff
@@ -0,0 +1,271 @@
+}
+
+// Service is an agent provisioner service that can be used by the provisioner
+// to retrieve container configuration for provisioning.
+type Service struct {
+	providerGetter func(context.Context) (Provider, error)
+	st             State
+}
+
+// State provides the service with access to controller/model config.
+type State interface {
+	// GetModelConfigKeyValues returns a model config object populated with
+	// values for the provided keys.
+	GetModelConfigKeyValues(context.Context, ...string) (map[string]string, error)
+	// ModelID returns the UUID of the current model.
+	ModelID(ctx context.Context) (model.UUID, error)
+}
+
+// NewService returns a new agent provisioner service.
+func NewService(
+	st State,
+	providerGetter providertracker.ProviderGetter[Provider],
+) *Service {
+	return &Service{
+		providerGetter: providerGetter,
+		st:             st,
+	}
+}
+
+// ContainerManagerConfigForType returns the container manager config for the
+// given container type.
+func (s *Service) ContainerManagerConfigForType(
+	ctx context.Context,
+	containerType instance.ContainerType,
+) (containermanager.Config, error) {
+	rval := containermanager.Config{}
+
+	modelID, err := s.st.ModelID(ctx)
+	if err != nil {
+		return rval, fmt.Errorf("getting ID for current model: %w", err)
+	}
+
+	cfg, err := s.st.GetModelConfigKeyValues(ctx,
+		config.LXDSnapChannel,
+		config.ContainerImageMetadataURLKey,
+		config.ContainerImageMetadataDefaultsDisabledKey,
+		config.ContainerImageStreamKey,
+	)
+	if err != nil {
+		return rval, fmt.Errorf(
+			"cannot get model config keys when calculating container manager config: %w",
+			err,
+		)
+	}
+
+	rval.ModelID = modelID
+	if containerType == instance.LXD {
+		rval.LXDSnapChannel = cfg[config.LXDSnapChannel]
+	}
+	rval.ImageMetadataURL = cfg[config.ContainerImageMetadataURLKey]
+	rval.MetadataDefaultsDisabled, _ = strconv.ParseBool(cfg[config.ContainerImageMetadataDefaultsDisabledKey])
+	rval.ImageStream = cfg[config.ContainerImageStreamKey]
+
+	return rval, nil
+}
+
+// ContainerNetworkingMethod determines the container networking method that
+// should be used, based on the model config key "container-networking-method"
+// and the current provider.
+func (s *Service) ContainerNetworkingMethod(ctx context.Context) (containermanager.NetworkingMethod, error) {
+	cfg, err := s.st.GetModelConfigKeyValues(ctx, config.ContainerNetworkingMethod)
+	if err != nil {
+		return "", fmt.Errorf("getting container networking method from model config: %w", err)
+	}
+
+	method := containermanager.NetworkingMethod(cfg[config.ContainerNetworkingMethod])
+	switch method {
+	case containermanager.NetworkingMethodLocal, containermanager.NetworkingMethodProvider:
+		return method, nil
+	case "":
+		// Auto-configure container networking method below
+	default:
+		return "", fmt.Errorf("invalid container networking method %q in model config", method)
+	}
```

Comment:
@tlm would we have a separate package with the constants for the model config enum, or just use string literals like this?
```suggestion
	switch cfg[config.ContainerNetworkingMethod] {
	case "local":
	        return containermanager.NetworkingMethodLocal, nil
	case "provider":
		return containermanager.NetworkingMethodProvider, nil
	case "":
		// Auto-configure container networking method below
	default:
		return "", fmt.Errorf("invalid container networking method %q in model config", method)
	}
```

Reply:
Um we I think let's define a constant for it. We have written a validator I believe to check the value on the way in so it makes sense to define enums.